// @nuvemerudita/react-controls/src/components/dropDownField/IDropDownFieldProps.ts
import { IOption } from './IOption';

// Interface for DropdownField props
export interface IDropdownFieldProps {
    options: IOption[]; // Required: Array of options for the dropdown
    onChange?: (selectedValue: string) => void; // Optional: Callback function to handle selection changes
    renderItem?: (item: IOption) => JSX.Element; // Required: Function to render each item in the dropdown
    validate?: (value: string) => string; // Optional: Function for validating the selected value
    label?: string | JSX.Element; // Optional: Label for the dropdown field
    placeholder?: string; // Optional: Placeholder text for the dropdown
    message?: string; // Optional: Custom message to display below the dropdown
    messageType?: "error" | "info" | "warning"; // Optional: Type of message to display
    required?: boolean; // Optional: Marks the field as required
    size?: "small" | "medium" | "large"; // Optional: Size of the Field's label
    hint?: string; // Optional: Additional hint text below the field
  }
