// @nuvemerudita/react-controls/src/components/renderHeader/useRenderHeaderStyles.ts
import {
  makeStyles,
  shorthands,
  tokens,
} from '@fluentui/react-components';

export const useRenderHeaderStyles = makeStyles({
  closeButton: {
    marginLeft: "auto",
    position: "absolute",
    top: "10px",
    right: "10px",
    zIndex: 99999,
  },

  divider: {
    width: "100%",
    height: "1px",
    backgroundColor: tokens.colorNeutralStroke1,
    marginTop: "6px",
  },

  renderHeaderContent: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "flex-start",
    alignItems: "stretch",

    width: "100%",
  },
  renderHeaderHeader: {
   

    display: "flex",
    flexDirection: "column",
    justifyContent: "flex-start",
    ...shorthands.padding("10px"),
    gap: "20px",
  },
  renderHeaderFooter: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "flex-start",
    ...shorthands.padding("20px"),
    gap: "20px",
  },
  renderHeaderBody: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "flex-start",
    ...shorthands.padding("20px"),
    gap: "20px",
  },
  renderHeaderTitleContainer: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "flex-start",
    alignItems: "center",
  },
  renderHeaderTitle: {
    display: "-webkit-box",
    "-webkit-line-clamp": "1",
    "-webkit-box-orient": "vertical",

    textAlign: "start",
    textOverflow: "ellipsis",
    paddingTop: "10px",
    wordBreak: "break-word",
  },
  renderHeaderDescription: {
    display: "-webkit-box",
    "-webkit-line-clamp": "4",
    "-webkit-box-orient": "vertical",

    textAlign: "start",
    textOverflow: "ellipsis",
    wordBreak: "break-word",
  },
  dialogTitleAndDescriptionContainer: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "start",
    alignItems: "start",
    paddingLeft: "10px",
    paddingRight: "20px",
  },
});
