// @nuvemerudita/react-controls/src/components/Carousel/useCarouselStyles.ts
/* eslint-disable @typescript-eslint/explicit-function-return-type */

import * as React from 'react';

 import { css } from '@emotion/css';

import { tokens } from '@fluentui/react-components';

 
import { ICarouselProps } from './ICarouselProps';
import { useComponentUtils } from '../../utils';

export const useCarouselStyles = (_props: ICarouselProps, containerWidth: number, ) => {
  const { getCurrentDevice } = useComponentUtils();

  const slideSizeBasedOnWidth = React.useCallback(() => {
    const deviceSize = getCurrentDevice(containerWidth);
    let slideSize = "95%";
    switch (deviceSize) {
      case "Medium":
        slideSize = "45%";
        break;
      case "Large":
        slideSize = "calc(95% / 4)";
        break;
      case "ExtraLarge":
        slideSize = "calc(95% / 5)";
        break;
      default:
        slideSize = "95%";
        break;
    }
    return slideSize;
  }, [containerWidth, getCurrentDevice]);

  const getSlideSpacing = React.useCallback(() => {
    const deviceSize = getCurrentDevice(containerWidth);

    let slideSpacing =  tokens.spacingVerticalM;

    switch (deviceSize) {
      case "Medium":
        slideSpacing =  tokens.spacingVerticalM;
        break;
      case "Large":
        slideSpacing =  tokens.spacingVerticalL;
        break;
        case "ExtraLarge":
        slideSpacing =  tokens.spacingVerticalXL;
        break;
      default:
        slideSpacing =  tokens.spacingVerticalM;
        break;
    }
    return slideSpacing;
  }, [containerWidth, getCurrentDevice]);

  const enableControls = css({
    display: "flex",
    justifyContent: "space-between",
    alignItems: "stretch",
    gap: 10,
    padding: 5,
  });

  const embla__viewport = css({
    overflow: "hidden",
    paddingTop: "10px",
    paddingBottom: "20px",
  });

  const embla__container = css({
    backfaceVisibility: "hidden",
    display: "flex",
    touchAction: "pan-y",
    
  });

  const embla = css({
    ".embla": {},
  });

  const embla__slide = css({
    minWidth: 0,
    flex: `0 0 ${slideSizeBasedOnWidth()}`,
    paddingLeft: `${getSlideSpacing()}`,
  });

const embla__buttons = css({
    display: "grid",
    gridTemplateColumns: "repeat(2, 1fr)",
    gap: 0,
    alignItems: "center",
});


  return { enableControls, embla__viewport, embla__container, embla, embla__slide, embla__buttons };
};
