// @nuvemerudita/react-controls/src/components/breadcrumb/useBreadcrumStyles.ts
import { IBreadcrumbItem } from './IBreadcrumItem';
import { css } from '@emotion/css';
import { tokens } from '@fluentui/react-components';

export const useBreadcrumStyles = () => {
  const styles = {
    container: css({
      display: 'flex',
      alignItems: 'center',
      alignContent: 'center',
      overflow: 'hidden',
      whiteSpace: 'nowrap',
      padding: '8px',
      backgroundColor: 'transparent'
    }),
    breadcrumbItem: css({
      minWidth: 'fit-content',
      padding: 0,
      margin: '0px',
    }),
    separator: css({
      marginRight: '8px',
    }),
  };
  const getStyles = (
    props: IBreadcrumbItem,
    size: 'small' | 'medium' | 'large' | 'current'
  ) => {
    const { icon } = props;
    switch (size) {
      case 'small':
        return css({
          fontSize: '12px',
          padding: '4px',
          paddingLeft: icon ? '0px' : '4px',
        });
      case 'medium':
        return css({
          fontSize: '14px',
          padding: '6px',
          paddingLeft: icon ? '0px' : '6px',
        });
      case 'large':
        return css({
          fontSize: '16px',
          padding: '8px',
          paddingLeft: icon ? '0px' : '8px',
        });
      case 'current':
        return css({
          fontWeight: tokens.fontWeightBold,
          cursor: 'default',
          display: '-webkit-box',
          '-webkit-line-clamp': '1',
          '-webkit-box-orient': 'vertical',
          overflow: 'hidden',
          textAlign: 'start',
          textOverflow: 'ellipsis',
          wordBreak: 'break-word',
          fontFamily: tokens.fontFamilyBase,
        });

      default:
        return css({
          fontSize: '14px',
          padding: '6px',
          paddingLeft: icon ? '0px' : '6px',
        });
    }
  };
  return { styles, getStyles };
};
