// @nuvemerudita/react-controls/src/components/renderDialog/RenderDialog.tsx
import * as React from 'react';

import {
  Dialog,
  DialogActions,
  DialogBody,
  DialogContent,
  DialogSurface,
  DialogTitle,
  makeStyles,
  mergeClasses,
} from '@fluentui/react-components';

export interface IRenderDialogProps {
  isOpen: boolean;
  dialogTitle?: string | React.ReactNode;
  dialogActions?: JSX.Element;
  onDismiss?: (open?: boolean) => void;
  minWidth?: number | string;
  maxWidth?: number | string;
  className?: string;
  minHeight?: number | string;
  maxHeight?: number | string;
  children: React.ReactNode;
}

const DEFAULT_MIN_WIDTH = 200;

const DEFAULT_MIN_HEIGHT = 200;

const useStyles = makeStyles({
  dialog: {
    width: "100%",
    height: "100%",
    overflow: "hidden",
    maxHeight: "600px",
    padding: 0,
  },
  dialogBody: {
    height: "calc(100% - 200px)",
  },
});

export const RenderDialog: React.FunctionComponent<IRenderDialogProps> = (
  props: React.PropsWithChildren<IRenderDialogProps>
) => {
  const { isOpen, dialogTitle, dialogActions, children, maxWidth, className, minHeight, minWidth, maxHeight } = props;
  const styles = useStyles();
  if (!isOpen) return <></>;
  return (
    <Dialog open={isOpen} modalType="modal">
      <DialogSurface
        className={mergeClasses(styles.dialog, className)}
        style={{
          maxWidth: maxWidth,
          minWidth: minWidth ?? DEFAULT_MIN_WIDTH,
          minHeight: minHeight ?? DEFAULT_MIN_HEIGHT,
          height: "fit-content",
          maxHeight: maxHeight ?? "",
        }}
      >
        <DialogTitle>{dialogTitle}</DialogTitle>
        <DialogBody className={styles.dialogBody}>
          <DialogContent>{children}</DialogContent>
        </DialogBody>
        <DialogActions fluid position="end">
          {dialogActions}
        </DialogActions>
      </DialogSurface>
    </Dialog>
  );
};
