// @nuvemerudita/react-controls/src/components/Center/useCenterStyles.tsx
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import * as React from 'react';

 import { css } from '@emotion/css';
 
import { ICenterProps } from './ICenterProps';
import { useComponentUtils } from '../../utils';

export const useCenterStyles = (props: ICenterProps) => {
  const { styles } = props;

  const { getBaseStyles } = useComponentUtils();
  const centerStyles = React.useCallback(
    (containerWidth: number, containerHeight: number): React.CSSProperties => {
       
      const baseStyles = getBaseStyles(props, containerWidth, containerHeight);
      return {
        ...baseStyles,
        ...styles,
         overflow: "hidden",
        display: "flex",
       
        justifyContent: "center",
        alignItems: "center",
        width: "100%",
        height: "100%",
       
      };
    },
    [getBaseStyles, props, styles]
  );

  const childrenStyles = css({
    justifyContent: "center",
    textAlign: "center",
  });

  return { centerStyles, childrenStyles };
};
