// @microsoft/sp-application-base/lib-commonjs/common/Killswitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFindAppManifestInsteadUsingFirstKSActivated = exports.isAddEngagementLogKSActivated = exports.isSPAppBarHiddenOnPublicationsKSActivated = exports.isOverrideWorkloadKSActivated = exports.isMoveResetBeforeUpdatePrefetchKSActive = exports.isBrokerLoginChromelessKSActivated = exports.isChromelessRenderKSActivated = exports.Killswitches = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
var Killswitches = /** @class */ (function () {
    function Killswitches() {
    }
    Killswitches.isDuplicateErrorMessageV2KillSwitchActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated(sp_core_library_1.Guid.parse('efa9490b-c037-465e-ae95-6eccef6061e8'), '5/11/20', 'Prevent duplicate error messages from appearing');
    };
    Killswitches.isChangeWorkloadForTeamsWorkBenchKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated('66f65dfa-6988-413b-a46d-6872886d05fb'
        /* '11/18/2021', 'Change Workload For Teams WorkBench' */
        );
    };
    Killswitches.isPageTypeForLayoutsKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated('8b8d7a10-d7ba-47f5-8ed5-1aa72caf73ac'
        /* '12/06/2021', 'Extract _layouts name for PageType logging' */
        );
    };
    Killswitches.isDisableNavigationCacheKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated('1f9b5b68-e4fa-4ca9-94df-b97b1af022f4'
        /* '04/22/2022', 'Disable Navigation Cache' */
        );
    };
    Killswitches.isAddEduClassToEngagementKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated('6529d10e-5177-40ef-9c95-4b3b947e8795'
        /* '05/06/2022',
        'Add edu class to engagement'*/
        );
    };
    Killswitches.isOverrideApplicationCustomizerDynamicDataSourceIdKSActivated = function () {
        return sp_core_library_1._SPKillSwitch.isActivated('6529d10e-5177-40ef-9c95-4b3b947e8795'
        /* '09/15/2022',
        'Override BaseApplicationCustomizer _dynamicDataSourceId method exclude instanceId.'*/
        );
    };
    return Killswitches;
}());
exports.Killswitches = Killswitches;
function isChromelessRenderKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('3007279f-0df2-4ac9-a11e-e067d3be1058'
    /* 10/13/2022,
    'allow chromeless apps to have onRender called' */
    );
}
exports.isChromelessRenderKSActivated = isChromelessRenderKSActivated;
function isBrokerLoginChromelessKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('3c191136-63f6-40dd-8e60-b170181959b9'
    /* 10/13/2022,
    'run broker login as a chromeless app' */
    );
}
exports.isBrokerLoginChromelessKSActivated = isBrokerLoginChromelessKSActivated;
function isMoveResetBeforeUpdatePrefetchKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('2277e093-d6f0-46de-af44-7924b6ffc048'
    /* '06/26/2023', 'Reset cache manager before update event' */
    );
}
exports.isMoveResetBeforeUpdatePrefetchKSActive = isMoveResetBeforeUpdatePrefetchKSActive;
/* Cross project Kill-switch. Being used in LinkInterceptor as well. */
function isOverrideWorkloadKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('da8015f3-c5fe-45f3-9220-ca73eb5cf979'
    /* '11/15/2022', 'Override workload KS' */
    );
}
exports.isOverrideWorkloadKSActivated = isOverrideWorkloadKSActivated;
// Cross-project KS
function isSPAppBarHiddenOnPublicationsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c99f5bdc-a1ed-4132-8569-966cd81dfcb2'
    /* '07/19/2023', 'Hides spAppBar on Campaign Publication scenarios' */
    );
}
exports.isSPAppBarHiddenOnPublicationsKSActivated = isSPAppBarHiddenOnPublicationsKSActivated;
function isAddEngagementLogKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('21a3705e-d0c2-48bd-b3aa-6e2cf41b53b3'
    /* '05/17/2024', 'Add Engagement Log' */
    );
}
exports.isAddEngagementLogKSActivated = isAddEngagementLogKSActivated;
function isFindAppManifestInsteadUsingFirstKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f298c6ca-1f18-4a32-bd13-dc5ed406ec66'
    /*'08/21/2024', 'aterentiev - Find app manifest instead of using first' */
    );
}
exports.isFindAppManifestInsteadUsingFirstKSActivated = isFindAppManifestInsteadUsingFirstKSActivated;
