// @fluentui/react-migration-v8-v9/lib-commonjs/components/Theme/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    black: function() {
        return _themeDuplicates.black;
    },
    blackAlpha: function() {
        return _themeDuplicates.blackAlpha;
    },
    brandWeb: function() {
        return _themeDuplicates.brandWeb;
    },
    createBrandVariants: function() {
        return _v9BrandVariantsShim.createBrandVariants;
    },
    createV8Theme: function() {
        return _v8ThemeShim.createV8Theme;
    },
    createV9Theme: function() {
        return _v9ThemeShim.createV9Theme;
    },
    grey: function() {
        return _themeDuplicates.grey;
    },
    grey10Alpha: function() {
        return _themeDuplicates.grey10Alpha;
    },
    grey12Alpha: function() {
        return _themeDuplicates.grey12Alpha;
    },
    grey14Alpha: function() {
        return _themeDuplicates.grey14Alpha;
    },
    hcButtonFace: function() {
        return _themeDuplicates.hcButtonFace;
    },
    hcButtonText: function() {
        return _themeDuplicates.hcButtonText;
    },
    hcCanvas: function() {
        return _themeDuplicates.hcCanvas;
    },
    hcCanvasText: function() {
        return _themeDuplicates.hcCanvasText;
    },
    hcDisabled: function() {
        return _themeDuplicates.hcDisabled;
    },
    hcHighlight: function() {
        return _themeDuplicates.hcHighlight;
    },
    hcHighlightText: function() {
        return _themeDuplicates.hcHighlightText;
    },
    hcHyperlink: function() {
        return _themeDuplicates.hcHyperlink;
    },
    sharedColors: function() {
        return _themeDuplicates.sharedColors;
    },
    white: function() {
        return _themeDuplicates.white;
    },
    whiteAlpha: function() {
        return _themeDuplicates.whiteAlpha;
    }
});
const _themeDuplicates = require("./themeDuplicates");
const _v8ThemeShim = require("./v8ThemeShim");
const _v9BrandVariantsShim = require("./v9BrandVariantsShim");
const _v9ThemeShim = require("./v9ThemeShim");
