// @fluentui/react-migration-v8-v9/lib-commonjs/components/Button/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ActionButtonShim: function() {
        return _ActionButtonShim.ActionButtonShim;
    },
    ButtonShim: function() {
        return _ButtonShim.ButtonShim;
    },
    CommandButtonShim: function() {
        return _CommandButtonShim.CommandButtonShim;
    },
    CompoundButtonShim: function() {
        return _CompoundButtonShim.CompoundButtonShim;
    },
    DefaultButtonShim: function() {
        return _DefaultButtonShim.DefaultButtonShim;
    },
    MenuButtonShim: function() {
        return _MenuButtonShim.MenuButtonShim;
    },
    PrimaryButtonShim: function() {
        return _PrimaryButtonShim.PrimaryButtonShim;
    },
    ToggleButtonShim: function() {
        return _ToggleButtonShim.ToggleButtonShim;
    },
    shimButtonProps: function() {
        return _shimButtonProps.shimButtonProps;
    }
});
const _ActionButtonShim = require("./ActionButtonShim");
const _ButtonShim = require("./ButtonShim");
const _CommandButtonShim = require("./CommandButtonShim");
const _CompoundButtonShim = require("./CompoundButtonShim");
const _DefaultButtonShim = require("./DefaultButtonShim");
const _MenuButtonShim = require("./MenuButtonShim");
const _PrimaryButtonShim = require("./PrimaryButtonShim");
const _shimButtonProps = require("./shimButtonProps");
const _ToggleButtonShim = require("./ToggleButtonShim");
