// @nuvemerudita/react-controls/src/components/renderDrawer/RenderDrawer.tsx
import * as React from 'react';

import {
  Drawer,
  DrawerBody,
  DrawerFooter,
  DrawerHeader,
  DrawerHeaderTitle,
  DrawerProps,
} from '@fluentui/react-components/unstable';

type DrawerType = Required<DrawerProps>["type"];

export interface IRenderDrawerProps {
  isOpen: boolean;
  onOpenChange?: (open: boolean) => void;
  title?: string | React.ReactNode;
  headerActions?: JSX.Element;
  footerActions?: JSX.Element;
  size?: "small" | "medium" | "large" | "full";
  separator?: boolean;
  type?: DrawerType;
  style?: React.CSSProperties;
  children: React.ReactNode;
  position?: "start" | "end";
  classNameHeader?: string;
  classNameBody?: string;
  classNameFooter?: string;
}

export const RenderDrawer: React.FunctionComponent<IRenderDrawerProps> = (
  props: React.PropsWithChildren<IRenderDrawerProps>
) => {
  const {
    isOpen,
    onOpenChange,
    title,
    headerActions,
    footerActions,
    size,
    separator,
    type,
    style,
    children,
    position,
    classNameHeader = "",
    classNameBody = "",
    classNameFooter = "",
  } = props;
  return (
    <>
      <Drawer
        type={type}
        separator={separator}
        open={isOpen}
        onOpenChange={(ev: React.SyntheticEvent<HTMLElement>, data: { open: boolean }) => {
          if (onOpenChange) {
            onOpenChange(data.open);
          }
        }}
        size={size}
        style={style}
        position={position}
      >
        <DrawerHeader className={classNameHeader}>
          <DrawerHeaderTitle style={{ width: "100%" }} action={headerActions}>
            {title}
          </DrawerHeaderTitle>
        </DrawerHeader>
        <DrawerBody className={classNameBody}>{children}</DrawerBody>
        <DrawerFooter className={classNameFooter}>{footerActions}</DrawerFooter>
      </Drawer>
    </>
  );
};
